prompt_poet/gcs_dict_template_loader.py

```diff
@@ -0,0 +1,87 @@
+import logging
+
+from jinja2 import BaseLoader
+from google.cloud import storage
+import os
+import jinja2 as j2
+
+from template_loaders import parse_template_path, TemplateLoader
+
+logger = logging.getLogger(__name__)
+
+
+class CacheLoader(BaseLoader):
+    """CacheLoader for Jinja2."""
+    def __init__(self, mapping):
+        self.mapping = mapping
+
+    def get_source(self, environment, template):
+        def uptodate():
+            return True
+
+        if template not in self.mapping:
+            raise j2.TemplateNotFound(template)
+
+        return self.mapping[template], template, uptodate
+
+
+class GCSDictTemplateLoader(TemplateLoader):
+    def __init__(self, bucket_name: str, template_path: str, gcs_client: storage.Client):
+        self._bucket_name = bucket_name
+        self._gcs_client = gcs_client
+        self._template_dir, self._template_name = parse_template_path(template_path)
+        self._mapping = {}
+        self._generation_data = {}
+
+    def _is_stale(self, blob: storage.Blob) -> bool:
+        """Check if the file needs to be downloaded."""
+        local_generation = self._generation_data.get(blob.name)
+        return local_generation != blob.generation
+
+    def _download(self):
+        """Download all files from the specified GCS directory to the local cache."""
+        bucket = self._gcs_client.bucket(self._bucket_name)
+        blobs = bucket.list_blobs(prefix=self._template_dir)
+        for blob in blobs:
+            if blob.name.endswith("/"):
+                continue
+            if not is_yaml_jinja(blob.name):
+                continue
```

Comment:
can you warn here that non yaml/jinja files should not be in the directory...so that if there is a problem with the is_yaml_jinja implementation or otherwise it is obvious to the user